### PR TITLE
Improve source extraction logic for coverage report generation

### LINKED ISSUE #30093: Incorrect code coverage reports are generated for native code
**Description:**
This is related to the code coverage reports generated for the [WebSocket module](https://github.com/ballerina-platform/module-ballerina-websocket).

Please refer to the following report.
https://app.codecov.io/gh/ballerina-platform/module-ballerina-websocket

In those reports, I could see couple of issues like,
1. Shows method signatures are not covered while the body of the method is covered.
image
image

2. Shows part of a code line is covered when the code line is separated into more than one line because the line exceeds the limit of 120.
image

3. Shows commented lines are not covered.
image
image

4. Shows empty lines are not covered
image

5. Shows imports are not covered
image

Can this be resolved as this has an impact on the final code coverage percentage?

### PR DESCRIPTION
**Pull request by Dilhasha**
## Purpose

Improve source extraction logic for coverage report generation. Fixes bug on duplicated code.
Fixes #30093 

## Approach
Improve source extraction logic for coverage report generation. Fixes bug on duplicated code.
Along with these changes, it's important to fine tune the includes provided when running tests with coverage.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by hemikak — DISMISSED** [review]

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java [hunk 1]
```diff
@@ -110,30 +111,29 @@ public CoverageReport(Module module, Map<String, ModuleCoverage> moduleCoverageM
...
         String packageName = this.module.packageInstance().packageName().toString();
-        String version = this.module.packageInstance().packageVersion().toString();
 
         List<Path> filteredPathList;
 
         if (!module.testDocumentIds().isEmpty()) {
             filteredPathList =
                     filterPaths(jBallerinaBackend.jarResolver().getJarFilePathsRequiredForTestExecution(
-                            this.module.moduleName()), jBallerinaBackend);
+                            this.module.moduleName()), jBallerinaBackend, module.packageInstance());
```
Comment (azinneera) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:120]: Shall we pass module to the filterPaths and refer to the `moduleName` and `packageInstance` in the method body?

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java [hunk 2]
```diff
@@ -110,30 +111,29 @@ public CoverageReport(Module module, Map<String, ModuleCoverage> moduleCoverageM
...
-            final CoverageBuilder coverageBuilder = analyzeStructure();
-            // Create Testerina coverage report
-            createReport(coverageBuilder.getBundle(title), moduleCoverageMap);
-            // Add additional dependency jars for Coverage XML if included
+            CoverageBuilder coverageBuilder = generateTesterinaCoverageReport(orgName, packageName, filteredPathList);
+            // Add additional dependency jars for Jacoco Coverage XML if included
             if (includesInCoverage != null) {
-                addCompiledSources(getDependencyJarList(jBallerinaBackend), orgName, packageName, version,
-                       true, includesInCoverage);
+                List<Path> dependencyPathList = getDependencyJarList(jBallerinaBackend);
```
Comment (azinneera) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:129]: Shall we move getting the jar list to a separate method?

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java [hunk 3]
```diff
@@ -382,15 +422,48 @@ private Document getDocumentFromModule(Module moduleInstance, String sourceFileN
...
         exclusionPathList.add(jBallerinaBackend.runtimeLibrary().path());
         for (JarLibrary library : ProjectUtils.testDependencies()) {
             exclusionPathList.add(library.path());
         }
+        // Add platform libs of this package to exclusion jar list
+        exclusionPathList.addAll(getPlatformLibsList(jBallerinaBackend, pkg));
         return exclusionPathList;
     }
 
+    private List<Path> getCodeGeneratedJarList(JBallerinaBackend jBallerinaBackend) {
```
Comment (azinneera) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:436]: Looks like there is no usage for this method.

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/CodeCoverageUtils.java [hunk 4]
```diff
@@ -81,14 +78,12 @@
      * @param destination path to extract the classes
      * @param orgName     org name of the project being executed
      * @param moduleName  name of the module being executed
-     * @param version     version of the module being executed
      * @throws NoSuchFileException if source file doesnt exist
      */
-    public static void unzipCompiledSource(Path source, Path destination, String orgName, String moduleName,
-                                           String version, boolean enableIncludesFilter,
```
Comment (azinneera) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/CodeCoverageUtils.java:88]: Don't we need the version also?

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java (hunk 1 above)
Comment (Dilhasha) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:120]: The moduleName is not passed to the filterPaths method. It's being used to get the required jar paths from JarResolver. So, I guess it's ok to keep this as is. WDYT?

Comment (azinneera) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:120]: Ack. +1

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/CodeCoverageUtils.java (hunk 4 above)
Comment (Dilhasha) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/CodeCoverageUtils.java:88]: It was used previously to create a structure and copy to bin path. Now it's not required.

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java (hunk 2 above)
Comment (Dilhasha) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java]: Fixed with 1fabfc7e25ef44de12795d3e47d8a39b0d2efede

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java (hunk 3 above)
Comment (Dilhasha) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java]: Fixed with 1fabfc7e25ef44de12795d3e47d8a39b0d2efede

**Review by azinneera — APPROVED** [review]

**Pull request merged**